src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -3802,6 +3802,7 @@ private bool ConnectAsync(SocketAsyncEventArgs e, bool userSocket)
 
                 WildcardBindForConnectIfNecessary(endPointSnapshot.AddressFamily);
 
+                bool canUseConnectEx = CanUseConnectEx(endPointSnapshot);
```

Comment:
Why does this need to be tested here vs inside of the `if` below?

Reply:
The check [will return false positive for UDS](https://github.com/dotnet/runtime/pull/787#issuecomment-592149956) if done later.

Maybe there is a smarter way to guard against this, but not sure if it's not worth the efforts.

Reply:
Definitely worth a comment though.

Reply:
> Why does this need to be tested here vs inside of the if below?

I had first put it in the `if` below, but it didn't work. Then Anton figured out the check should be put higher. But when I look at the code deeper I don't understand why.

`CanUseConnectEx` is implemented as follows:
```
        private bool CanUseConnectEx(EndPoint remoteEP)
        {
            return (_socketType == SocketType.Stream) &&
                (_rightEndPoint != null || remoteEP.GetType() == typeof(IPEndPoint));
        }
```

We want this to return `true` for stream sockets, and `false` for non-stream sockets.
The latter is ok. The former is `ok` when we connect to an IPEndPoint.
When we connect to a DNSEndpoint, `_rightEndpoint` should not be `null` for this to be ok.
But actually moving it sooner should cause it to be `null`.
I don't understand why this works, I'm missing something.

I will put it lower and we can see what tests fail.

Reply:
@tmds I'm guilty not naming the test. It will be `Socket_ConnectAsyncUnixDomainSocketEndPoint_NotServer`.

> But actually moving it sooner should cause it to be null.

The reason is that `WildcardBindForConnectIfNecessary` will not bind `_rightEndpoint` for UDS. I think the behavior is the same on Linux, but it doesn't lead to error, since `DoOperationConnectEx` is a wrapper for `DoOperationConnect` on Unix.

Reply:
@antonfirsov am I misreading the code? `CanUseConnectEx` is being called on `endPointSnapshot`, which doesn't appear to be modified between the call site and where the bool is tested.

Reply:
`Socket_ConnectAsyncUnixDomainSocketEndPoint_Success` fails.
For this test we are in the `SocketType.Stream` case, and using a `UnixDomainSocketEndPoint`. We need `CanUseConnectEx` to return `false` because `ConnectEx` doesn't support this end-point type.
Probably the proper thing to do is to add a type check for `UnixDomainSocketEndPoint` too?

Reply:
> Probably the proper thing to do is to add a type check for UnixDomainSocketEndPoint?

That, or `CanUseConnectEx` should be called when `_rightEndPoint = null`.
It's not clear why there is a `_rightEndPoint` check in `CanUseConnectEx` for me. Any thoughts @scalablecory @antonfirsov ?

Reply:
Ah, that is what I'm missing -- the `_rightEndPoint` check.

> It's not clear why there is a `_rightEndPoint` check in `CanUseConnectEx` for me.

`_rightEndPoint` would only be set if the socket is already bound; I'm not sure why you would want to bypass the remote endpoint check if the socket is already bound. Maybe the expectation is that the OS layer will fail it in another way.

Reply:
It may be an optimization assuming `WildcardBindForConnectIfNecessary` will bind the socket and create a `_rightEndPoint`.

Reply:
> `CanUseConnectEx` is being called on `endPointSnapshot`, which doesn't appear to be modified between the call site and where the bool is tested.

@scalablecory `CanUseConnectEx()` also depends on `_rightEndpoint`, which changed as the socket becomes bound:
https://github.com/dotnet/runtime/blob/08ce33ad7a66b84604afa2b1145983adec72b08c/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3808-L3811

> It's not clear why there is a `_rightEndPoint` check in `CanUseConnectEx` for me.

@tmds the check `_rightEndpoint != null` is equivalent to checking `this.IsBound`. My understanding is that `CanUseConnectEx` only works with sockets which are bound.

> Probably the proper thing to do is to add a type check for `UnixDomainSocketEndPoint`?

Sounds like the proper solution for this problem. I will try implementing it like this, and come back to you so you don't have to code blindly.

Reply:
> Probably the proper thing to do is to add a type check for UnixDomainSocketEndPoint?

I haven't followed the whole discussion, but there are other potential endpoint types.  If something can only work with a subset of them, we should explicitly check for the ones known to work rather than trying to screen out ones known to not work.

Reply:
@stephentoub this would be addressed with [my suggestion](https://github.com/antonfirsov/runtime/commit/5196b015cdfa30bb347c96c2381b531a02243370).

Reply:
```
return _socketType == SocketType.Stream && remoteEP is IPEndPoint;
```
If we can leave out `_rightEndpoint` from `CanUseConnectEx` that seems like a good thing to me. I also think it is some sort of optimization that may be left out.